File: experiments/problem1_nonperiodicity/analyze_period_two_phase_frontier_lift_recursion.cpp
#include <algorithm>
#include <array>
#include <bitset>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <map>
#include <optional>
#include <stdexcept>
#include <string>
#include <tuple>
#include <vector>

namespace {
constexpr int MAXIMUM_COMPLEXITY = 25;
constexpr std::uint64_t KNOWN_COUNTEREXAMPLE = 0x1bcd3a7b3fdfbULL;
constexpr std::array<unsigned, 4> CHILD_MASK{0b1011, 0b1100, 0b1110, 0b0011};

std::uint64_t forward_generator(char name, std::uint64_t state) {
    const std::uint64_t stepped = state ^ ((state << 1) | (state << 2));
    if (name == 't') return stepped;
    if (name == 'u') return stepped ^ 1;
    if (name == 'p') return stepped ^ 1 ^ ((state & 1) == 0 ? 2 : 0);
    throw std::invalid_argument("unknown generator");
}

int bit_length(std::uint64_t value) {
    return value == 0 ? 0 : 64 - __builtin_clzll(value);
}

int expected_bits(char phase, int complexity) {
    return phase == 'p' ? 2 * complexity : 2 * complexity - 1;
}

std::optional<std::uint64_t> inverse_t(std::uint64_t output) {
    if (output == 0) return std::uint64_t{0};
    const int width = bit_length(output) - 2;
    if (width <= 0) return std::nullopt;
    std::uint64_t state = 0;
    for (int position = 0; position < width; ++position) {
        unsigned lower = 0;
        if (position >= 1) lower |= static_cast<unsigned>((state >> (position - 1)) & 1);
        if (position >= 2) lower |= static_cast<unsigned>((state >> (position - 2)) & 1);
        const unsigned value = static_cast<unsigned>((output >> position) & 1) ^ lower;
        state |= static_cast<std::uint64_t>(value) << position;
    }
    if (forward_generator('t', state) != output) return std::nullopt;
    return state;
}

std::optional<std::uint64_t> inverse_generator(char name, std::uint64_t output) {
    if (name == 't') return inverse_t(output);
    std::optional<std::uint64_t> state;
    if (name == 'u') {
        state = inverse_t(output ^ 1);
    } else if (name == 'p') {
        const unsigned recovered_low = static_cast<unsigned>(output & 1) ^ 1;
        const std::uint64_t adjusted = output ^ 1 ^ (recovered_low == 0 ? 2 : 0);
        state = inverse_t(adjusted);
    } else {
        throw std::invalid_argument("unknown inverse generator");
    }
    if (!state || forward_generator(name, *state) != output) return std::nullopt;
    return state;
}

std::optional<std::uint64_t> candidate_parent(std::uint64_t quotient, int digit) {
    const char generator = digit == 0 ? 't' : digit == 1 ? 'u' : 'p';
    const auto residual = inverse_generator(generator, quotient);
    if (!residual) return std::nullopt;
    return 4 * (*residual) + static_cast<std::uint64_t>(digit);
}

std::vector<std::uint64_t> next_level(const std::vector<std::uint64_t>& current) {
    std::vector<std::uint64_t> next;
    next.reserve(current.size() * 3);
    for (std::uint64_t state : current) {
        for (char generator : {'t', 'u', 'p'}) next.push_back(forward_generator(generator, state));
    }
    std::sort(next.begin(), next.end());
    next.erase(std::unique(next.begin(), next.end()), next.end());
    return next;
}

bool contains(const std::vector<std::uint64_t>& values, std::uint64_t target) {
    return std::binary_search(values.begin(), values.end(), target);
}

unsigned actual_fiber(const std::vector<std::uint64_t>& next, std::uint64_t quotient) {
    unsigned mask = 0;
    for (int digit = 0; digit < 4; ++digit) {
        if (contains(next, 4 * quotient + static_cast<std::uint64_t>(digit))) mask |= 1U << digit;
    }
    return mask;
}

bool allowed_fiber(unsigned mask) {
    return mask == 0b0000 || mask == 0b0011 || mask == 0b1011
        || mask == 0b1100 || mask == 0b1111;
}

struct Totals {
    std::uint64_t outputs = 0;
    std::uint64_t quotients = 0;
    std::uint64_t candidate_checks = 0;
    std::array<std::uint64_t, 16> fibers{};
    std::array<std::uint64_t, 16> predecessors{};
    std::uint64_t checksum = 1469598103934665603ULL;
};

Totals run_phase(char phase) {
    std::vector<std::uint64_t> current{phase == 'p' ? 3ULL : 1ULL};
    Totals totals;
    std::cout << "phase=" << phase << '\n';
    for (int complexity = 1; complexity <= MAXIMUM_COMPLEXITY; ++complexity) {
        for (std::uint64_t state : current) {
            if (bit_length(state) != expected_bits(phase, complexity)) {
                throw std::runtime_error("frontier bit-length law failed");
            }
            totals.checksum ^= state + (static_cast<std::uint64_t>(complexity) << 56);
            totals.checksum *= 1099511628211ULL;
        }
        totals.outputs += current.size();
        if (complexity == MAXIMUM_COMPLEXITY) {
            std::cout << "level=" << complexity << " outputs=" << current.size() << '\n';
            break;
        }

        const std::vector<std::uint64_t> next = next_level(current);
        std::array<std::uint64_t, 16> level_fibers{};
        std::uint64_t uncovered = 0;
        for (std::uint64_t quotient : current) {
            unsigned predecessor = 0;
            unsigned predicted = 0;
            for (int digit = 0; digit < 4; ++digit) {
                const auto parent = candidate_parent(quotient, digit);
                if (parent && contains(current, *parent)) {
                    predecessor |= 1U << digit;
                    predicted |= CHILD_MASK[digit];
                }
                ++totals.candidate_checks;
            }
            if ((predecessor & 0b0100) && !(predecessor & 0b1000)) {
                throw std::runtime_error("digit-two predecessor lacked digit-three mate");
            }
            const unsigned actual = actual_fiber(next, quotient);
            if (predicted != actual) throw std::runtime_error("lift recursion mismatch");
            if (!allowed_fiber(actual)) throw std::runtime_error("fiber escaped alphabet");
            ++totals.fibers[actual];
            ++totals.predecessors[predecessor];
            ++level_fibers[actual];
            ++totals.quotients;
            uncovered += actual == 0;
        }
        std::cout << "level=" << complexity
                  << " outputs=" << current.size()
                  << " next_outputs=" << next.size()
                  << " uncovered=" << uncovered
                  << " masks_0011=" << level_fibers[0b0011]
                  << " masks_1011=" << level_fibers[0b1011]
                  << " masks_1100=" << level_fibers[0b1100]
                  << " masks_1111=" << level_fibers[0b1111] << '\n';
        current = next;
    }
    std::cout << "phase_outputs_checked=" << totals.outputs << '\n';
    std::cout << "phase_quotients_checked=" << totals.quotients << '\n';
    std::cout << "phase_candidate_parent_checks=" << totals.candidate_checks << '\n';
    for (unsigned mask : {0U, 3U, 11U, 12U, 15U}) {
        std::cout << "phase_fiber_" << std::bitset<4>(mask)
                  << '=' << totals.fibers[mask] << '\n';
    }
    std::cout << "phase_checksum=0x" << std::hex << std::setw(16)
              << std::setfill('0') << totals.checksum << std::dec
              << std::setfill(' ') << '\n';
    return totals;
}

using Key = std::tuple<char, int, std::uint64_t>;
std::map<Key, bool> member_cache;
std::map<Key, std::string> witness_cache;

bool member(char phase, int complexity, std::uint64_t target) {
    const Key key{phase, complexity, target};
    if (const auto it = member_cache.find(key); it != member_cache.end()) return it->second;
    bool result = false;
    if (complexity >= 1 && bit_length(target) == expected_bits(phase, complexity)) {
        if (complexity == 1) {
            result = target == (phase == 'p' ? 3ULL : 1ULL);
        } else {
            const std::uint64_t quotient = target >> 2;
            const int digit = static_cast<int>(target & 3);
            for (int parent_digit = 0; parent_digit < 4 && !result; ++parent_digit) {
                if (((CHILD_MASK[parent_digit] >> digit) & 1U) == 0) continue;
                const auto parent = candidate_parent(quotient, parent_digit);
                if (parent && member(phase, complexity - 1, *parent)) result = true;
            }
        }
    }
    member_cache.emplace(key, result);
    return result;
}

std::string witness(char phase, int complexity, std::uint64_t target) {
    const Key key{phase, complexity, target};
    if (const auto it = witness_cache.find(key); it != witness_cache.end()) return it->second;
    if (!member(phase, complexity, target)) return {};
    if (complexity == 1) {
        const std::string result(1, phase);
        witness_cache.emplace(key, result);
        return result;
    }
    const std::uint64_t quotient = target >> 2;
    const int digit = static_cast<int>(target & 3);
    for (int parent_digit = 0; parent_digit < 4; ++parent_digit) {
        if (((CHILD_MASK[parent_digit] >> digit) & 1U) == 0) continue;
        const auto parent = candidate_parent(quotient, parent_digit);
        if (!parent || !member(phase, complexity - 1, *parent)) continue;
        std::string prefix = witness(phase, complexity - 1, *parent);
        for (char generator : {'t', 'u', 'p'}) {
            if (forward_generator(generator, *parent) == target) {
                prefix.push_back(generator);
                witness_cache.emplace(key, prefix);
                return prefix;
            }
        }
    }
    throw std::runtime_error("membership lacked witness");
}

std::uint64_t apply_word(const std::string& word) {
    if (word.empty()) throw std::invalid_argument("empty word");
    std::uint64_t state = word.front() == 'p' ? 3ULL : 1ULL;
    for (std::size_t index = 1; index < word.size(); ++index) {
        state = forward_generator(word[index], state);
    }
    return state;
}
}  // namespace

int main() {
    const Totals p = run_phase('p');
    const Totals u = run_phase('u');
    if (!member('u', 25, KNOWN_COUNTEREXAMPLE)) {
        throw std::runtime_error("recursive criterion missed counterexample");
    }
    const std::string word = witness('u', 25, KNOWN_COUNTEREXAMPLE);
    if (word.size() != 25 || apply_word(word) != KNOWN_COUNTEREXAMPLE) {
        throw std::runtime_error("recursive witness replay failed");
    }
    std::cout << "known_counterexample=0x" << std::hex << KNOWN_COUNTEREXAMPLE << std::dec << '\n';
    std::cout << "recovered_generator_word=" << word << '\n';
    std::cout << "recursive_cache_entries=" << member_cache.size() << '\n';
    std::cout << "total_outputs_checked=" << p.outputs + u.outputs << '\n';
    std::cout << "total_quotients_checked=" << p.quotients + u.quotients << '\n';
    std::cout << "total_candidate_parent_checks=" << p.candidate_checks + u.candidate_checks << '\n';
    std::cout << "exact_lift_recursion=true\n";
    std::cout << "fiber_alphabet=0000,0011,1011,1100,1111\n";
    return 0;
}

File: experiments/problem1_nonperiodicity/analyze_period_two_phase_frontier_lift_recursion.py
#!/usr/bin/env python3
"""Exact recursive base-four lift criterion for ordinary phase frontiers.

For phase a in {p,u}, let O_(a,k) be the distinct ordinary outputs at
complexity k. If a target at level k+1 is y=4h+e, then its possible level-k
parents are determined by the low digit d of the parent and the partial inverse
of t, u, or p applied to h. This gives an exact recursive membership test that
does not enumerate the complete frontier.
"""
from __future__ import annotations

import argparse
import hashlib
import json
from collections import Counter
from functools import lru_cache
from typing import Any

PHASES = ("p", "u")
DEFAULT_MAXIMUM_COMPLEXITY = 16
ABSOLUTE_MAXIMUM_COMPLEXITY = 20
KNOWN_COUNTEREXAMPLE = 0x1BCD3A7B3FDFB
CHILD_DIGIT_MASK = {
    0: 0b1011,
    1: 0b1100,
    2: 0b1110,
    3: 0b0011,
}
ALLOWED_FIBER_MASKS = (0b0000, 0b0011, 0b1011, 0b1100, 0b1111)


class LiftRecursionLimitError(RuntimeError):
    """Raised before a controlled exhaustive campaign exceeds its cap."""


def forward_generator(name: str, state: int) -> int:
    if state < 0:
        raise ValueError("state must be nonnegative")
    stepped = state ^ ((state << 1) | (state << 2))
    if name == "t":
        return stepped
    if name == "u":
        return stepped ^ 1
    if name == "p":
        return stepped ^ 1 ^ (2 if state & 1 == 0 else 0)
    raise ValueError("unknown generator")


def frontier_children(state: int) -> tuple[int, ...]:
    return tuple(sorted({forward_generator(name, state) for name in "tup"}))


def phase_start(phase: str) -> int:
    if phase == "p":
        return 3
    if phase == "u":
        return 1
    raise ValueError("phase must be p or u")


def expected_bit_length(phase: str, complexity: int) -> int:
    if complexity <= 0:
        raise ValueError("complexity must be positive")
    return 2 * complexity if phase == "p" else 2 * complexity - 1


def inverse_t(output: int) -> int | None:
    """Return the unique exact nonnegative t-preimage, or None."""
    if output < 0:
        raise ValueError("output must be nonnegative")
    if output == 0:
        return 0
    width = output.bit_length() - 2
    if width <= 0:
        return None
    state = 0
    for position in range(width):
        lower = 0
        if position >= 1:
            lower |= (state >> (position - 1)) & 1
        if position >= 2:
            lower |= (state >> (position - 2)) & 1
        bit = ((output >> position) & 1) ^ lower
        state |= bit << position
    return state if forward_generator("t", state) == output else None


def inverse_generator(name: str, output: int) -> int | None:
    if output < 0:
        raise ValueError("output must be nonnegative")
    if name == "t":
        return inverse_t(output)
    if name == "u":
        state = inverse_t(output ^ 1)
    elif name == "p":
        recovered_low_bit = (output & 1) ^ 1
        adjusted = output ^ 1 ^ (2 if recovered_low_bit == 0 else 0)
        state = inverse_t(adjusted)
    else:
        raise ValueError("unknown generator")
    if state is None or forward_generator(name, state) != output:
        return None
    return state


def candidate_parent(quotient: int, parent_digit: int) -> int | None:
    if quotient < 0 or parent_digit not in range(4):
        raise ValueError("invalid candidate-parent argument")
    generator = "t" if parent_digit == 0 else "u" if parent_digit == 1 else "p"
    residual = inverse_generator(generator, quotient)
    return None if residual is None else 4 * residual + parent_digit


def predicted_fiber_mask(current: set[int], quotient: int) -> tuple[int, int]:
    predecessor_mask = 0
    fiber_mask = 0
    for digit in range(4):
        parent = candidate_parent(quotient, digit)
        if parent is not None and parent in current:
            predecessor_mask |= 1 << digit
            fiber_mask |= CHILD_DIGIT_MASK[digit]
    if predecessor_mask & 0b0100 and not predecessor_mask & 0b1000:
        raise AssertionError("residue-two predecessor lacked its residue-three mate")
    if fiber_mask not in ALLOWED_FIBER_MASKS:
        raise AssertionError("fiber escaped the five-mask alphabet")
    return predecessor_mask, fiber_mask


def actual_fiber_mask(next_level: set[int], quotient: int) -> int:
    return sum(1 << digit for digit in range(4) if 4 * quotient + digit in next_level)


def enumerate_phase(phase: str, maximum_complexity: int) -> dict[str, Any]:
    current = {phase_start(phase)}
    levels: list[dict[str, Any]] = []
    total_outputs = 0
    total_quotients = 0
    total_candidate_checks = 0
    total_mask_counts: Counter[int] = Counter()
    total_predecessor_counts: Counter[int] = Counter()

    for complexity in range(1, maximum_complexity + 1):
        expected = expected_bit_length(phase, complexity)
        if any(state.bit_length() != expected for state in current):
            raise AssertionError("frontier bit-length law failed")
        total_outputs += len(current)
        if complexity == maximum_complexity:
            levels.append({"complexity": complexity, "outputs": len(current)})
            break

        next_level = {child for state in current for child in frontier_children(state)}
        mask_counts: Counter[int] = Counter()
        predecessor_counts: Counter[int] = Counter()
        uncovered = 0
        for quotient in current:
            predecessor_mask, predicted = predicted_fiber_mask(current, quotient)
            actual = actual_fiber_mask(next_level, quotient)
            if predicted != actual:
                raise AssertionError("exact lift recursion failed")
            mask_counts[actual] += 1
            predecessor_counts[predecessor_mask] += 1
            uncovered += actual == 0
            total_quotients += 1
            total_candidate_checks += 4
        total_mask_counts.update(mask_counts)
        total_predecessor_counts.update(predecessor_counts)
        levels.append(
            {
                "complexity": complexity,
                "outputs": len(current),
                "next_outputs": len(next_level),
                "uncovered_quotients": uncovered,
                "fiber_masks": {
                    f"0b{mask:04b}": count for mask, count in sorted(mask_counts.items())
                },
            }
        )
        current = next_level

    return {
        "phase": phase,
        "levels": levels,
        "outputs_checked": total_outputs,
        "quotients_checked": total_quotients,
        "candidate_parent_checks": total_candidate_checks,
        "fiber_mask_totals": {
            f"0b{mask:04b}": count for mask, count in sorted(total_mask_counts.items())
        },
        "predecessor_mask_totals": {
            f"0b{mask:04b}": count
            for mask, count in sorted(total_predecessor_counts.items())
        },
    }


@lru_cache(maxsize=None)
def frontier_member(phase: str, complexity: int, target: int) -> bool:
    """Decide exact frontier membership by recursive lift inversion."""
    if complexity < 1 or target < 0:
        return False
    if target.bit_length() != expected_bit_length(phase, complexity):
        return False
    if complexity == 1:
        return target == phase_start(phase)

    quotient, digit = divmod(target, 4)
    for parent_digit in range(4):
        if not (CHILD_DIGIT_MASK[parent_digit] >> digit) & 1:
            continue
        parent = candidate_parent(quotient, parent_digit)
        if parent is not None and frontier_member(phase, complexity - 1, parent):
            return True
    return False


@lru_cache(maxsize=None)
def frontier_witness(phase: str, complexity: int, target: int) -> str | None:
    if not frontier_member(phase, complexity, target):
        return None
    if complexity == 1:
        return phase

    quotient, digit = divmod(target, 4)
    for parent_digit in range(4):
        if not (CHILD_DIGIT_MASK[parent_digit] >> digit) & 1:
            continue
        parent = candidate_parent(quotient, parent_digit)
        if parent is None or not frontier_member(phase, complexity - 1, parent):
            continue
        prefix = frontier_witness(phase, complexity - 1, parent)
        if prefix is None:
            continue
        for generator in "tup":
            if forward_generator(generator, parent) == target:
                return prefix + generator
    raise AssertionError("membership had no generator witness")


def apply_word(word: str) -> int:
    if not word or word[0] not in "pu":
        raise ValueError("word must begin with phase p or u")
    state = phase_start(word[0])
    for generator in word[1:]:
        state = forward_generator(generator, state)
    return state


def strict_examples() -> dict[str, Any]:
    examples = {
        "phase_p": {"phase": "p", "complexity": 2, "quotient": 12},
        "phase_u": {"phase": "u", "complexity": 3, "quotient": 26},
    }
    for row in examples.values():
        phase = row["phase"]
        complexity = row["complexity"]
        current = {phase_start(phase)}
        for _ in range(1, complexity):
            current = {child for state in current for child in frontier_children(state)}
        next_level = {child for state in current for child in frontier_children(state)}
        predecessor, predicted = predicted_fiber_mask(current, row["quotient"])
        actual = actual_fiber_mask(next_level, row["quotient"])
        if predicted != actual:
            raise AssertionError("strict example recursion mismatch")
        row["predecessor_mask"] = f"0b{predecessor:04b}"
        row["fiber_mask"] = f"0b{actual:04b}"
        row["lifts"] = [
            4 * row["quotient"] + digit
            for digit in range(4)
            if (actual >> digit) & 1
        ]
    return examples


def run_campaign(maximum_complexity: int = DEFAULT_MAXIMUM_COMPLEXITY) -> dict[str, Any]:
    if not 2 <= maximum_complexity <= ABSOLUTE_MAXIMUM_COMPLEXITY:
        raise LiftRecursionLimitError("maximum complexity outside controlled range")

    phase_rows = {
        phase: enumerate_phase(phase, maximum_complexity) for phase in PHASES
    }
    frontier_member.cache_clear()
    frontier_witness.cache_clear()
    if not frontier_member("u", 25, KNOWN_COUNTEREXAMPLE):
        raise AssertionError("recursive criterion missed known counterexample")
    witness = frontier_witness("u", 25, KNOWN_COUNTEREXAMPLE)
    if witness is None or len(witness) != 25 or apply_word(witness) != KNOWN_COUNTEREXAMPLE:
        raise AssertionError("recursive counterexample witness failed")

    payload: dict[str, Any] = {
        "status": "exact-recursive-membership-and-finite-exhaustion",
        "maximum_complexity": maximum_complexity,
        "theorem": {
            "membership": (
                "4h+e lies in O_(a,k+1) iff at least one exact candidate parent "
                "selected by the partial inverses of t,u,p lies in O_(a,k) and "
                "its parent-digit contribution contains e"
            ),
            "child_digit_masks": {
                str(digit): f"0b{mask:04b}"
                for digit, mask in CHILD_DIGIT_MASK.items()
            },
            "fiber_alphabet": [f"0b{mask:04b}" for mask in ALLOWED_FIBER_MASKS],
        },
        "phases": phase_rows,
        "strict_examples": strict_examples(),
        "known_counterexample": {
            "phase": "u",
            "complexity": 25,
            "state": KNOWN_COUNTEREXAMPLE,
            "state_hex": hex(KNOWN_COUNTEREXAMPLE),
            "recursive_membership": True,
            "recovered_generator_word": witness,
            "recursive_cache_entries": frontier_member.cache_info().currsize,
        },
        "scientific_boundary": (
            "The recursion decides ordinary phase-frontier membership exactly. It "
            "does not prove that actual return penalties recur positively, prove "
            "phase-complexity divergence, exclude eventual center period two, or "
            "solve Rule 30 center nonperiodicity."
        ),
    }
    encoded = json.dumps(payload, sort_keys=True, separators=(",", ":")).encode()
    payload["certificate_sha256"] = hashlib.sha256(encoded).hexdigest()
    return payload


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--maximum-complexity", type=int, default=DEFAULT_MAXIMUM_COMPLEXITY
    )
    args = parser.parse_args()
    print(json.dumps(run_campaign(args.maximum_complexity), indent=2, sort_keys=True))


if __name__ == "__main__":
    main()
